File: Go/day-02.go
// Ref: append(foo[:i], foo[i+1:]...) overwrites foo. https://stackoverflow.com/a/57213476/1200316

package main

import (
	"encoding/json" // Package for JSON encoding and decoding, used here for fast array parsing.
	"fmt"           // Package for formatted I/O (printing results).
	"os"            // Package for operating system interactions (reading files).
	"slices"        // Go 1.21+ package with optimized slice operations (cloning, deleting).
	"strings"       // Package for string manipulation (trimming, splitting, replacing).
)

// The main function is the entry point for the Go program execution.
func main() {
	// os.ReadFile reads the entire contents of "input.txt" into a byte slice.
	// The underscore (_) ignores the error returned by os.ReadFile for simplicity.
	input, _ := os.ReadFile("../inputs/input-02.txt")

	// Initialize counters for the two parts of the problem.
	part1, part2 := 0, 0

	// --- Input Processing Pipeline ---

	// string(input): Converts the byte slice into a string.
	// strings.TrimSpace(...): Removes leading/trailing whitespace (including newlines) from the file content.
	// strings.Split(..., "\n"): Splits the cleaned string into an array of individual lines (reports).
	// The loop iterates over each line (s) in the resulting slice of strings.
	for _, s := range strings.Split(strings.TrimSpace(string(input)), "\n") {

		// Variable to hold the parsed report numbers as a slice of integers.
		var report []int

		// --- Efficient String-to-Slice Parsing using JSON ---

		// strings.ReplaceAll(s, " ", ","): Replaces all spaces in the line with commas.
		//    (e.g., "1 2 3" becomes "1,2,3")
		// "[" + ... + "]": Wraps the comma-separated numbers in square brackets to form a valid JSON array string. (e.g., "[1,2,3]")
		// []byte(...): Converts the JSON string to a byte slice, required by Unmarshal.
		// json.Unmarshal(...): Parses the byte slice and efficiently populates the 'report' slice.
		json.Unmarshal([]byte("["+strings.ReplaceAll(s, " ", ",")+"]"), &report)

		// --- Part 1: Check original report safety ---
		if check(report) {
			part1++ // Increment if the original sequence is safe.
		}

		// --- Part 2: Check safety after removing one element ---
		// Iterate through every possible index 'i' to simulate deleting one element.
		for i := range report {
			// 1. slices.Clone(report): Creates a shallow copy of the report slice. This is essential
			//    because slices.Delete modifies the slice in place.
			// 2. slices.Delete(..., i, i+1): Creates a new slice with the element at index 'i' removed.
			//    The slice is passed to the 'check' function.
			if check(slices.Delete(slices.Clone(report), i, i+1)) {
				part2++ // Increment if a safe sub-sequence is found.
				break   // Stop checking this row and move to the next file line (since we only need ONE safe deletion).
			}
		}
	}

	// Print the final counts for Part 1 and Part 2.
	fmt.Println("Part 1 answer: ", part1)
	fmt.Println("Part 2 answer: ", part2)
}

// --- Safety Check Logic ---

// check verifies if the report sequence 'r' is safe according to the problem rules.
// A sequence is safe if the differences between consecutive numbers are:
// 1. Consistently positive OR consistently negative (no sign flips).
// 2. Have an absolute value between 1 and 3 (inclusive).
func check(r []int) bool {
	// A sequence with 0 or 1 element is trivially safe.
	if len(r) < 2 {
		return true
	}

	// Loop starts at the second element (i=1) to calculate the difference with the previous element (i-1).
	for i := 1; i < len(r); i++ {
		// Calculate the difference (d) between the current element and the previous one.
		d := r[i] - r[i-1]

		// Get the sign of the *first* difference (which dictates the required trend: increasing/decreasing).
		// This uses the first two elements of the *current* sequence being checked (r).
		firstDiffSign := r[1] - r[0]

		// --- Check 1: Sign Consistency ---
		// d * firstDiffSign <= 0: This is a fast way to check if the sign flipped or if d is zero.
		// If both signs are the same (d > 0 and firstDiffSign > 0, OR d < 0 and firstDiffSign < 0),
		// the product will be positive. If the trend flips, the product will be negative (<= 0).

		// --- Check 2: Magnitude Constraint ---
		// d < -3 || d > 3: Checks if the absolute difference is too large (> 3).

		if d*firstDiffSign <= 0 || d < -3 || d > 3 {
			return false // Immediately fail if any single difference violates the rules.
		}
	}

	// If the loop completes without finding any violations, the sequence is safe.
	return true
}

File: Py/day-02.py
def is_safe(row):
    """
    Checks if a sequence of numbers is 'safe'. A sequence is safe if the difference
    between every consecutive pair of numbers (the 'increments') falls entirely
    within the set {1, 2, 3} or entirely within the set {-1, -2, -3}.

    The function uses set operations for a concise check of all unique differences.

    :param row: A list of integers (the sequence).
    :return: True if the sequence is safe, False otherwise.
    """
    # Calculate the unique differences between each pair of consecutive elements in the list `row`.
    # Prodices a set so duplicates are removed.
    # It computes the difference between each pair of consecutive elements in `row`.
    inc = {row[i + 1] - row[i] for i in range(len(row) - 1)}

    # Check 1: Does the set of unique increments (inc) contain *only* values from {1, 2, 3}?
    # The '<=' operator checks if 'inc' is a subset of the target set.
    is_positive_safe = inc <= {1, 2, 3}

    # Check 2: Does the set of unique increments (inc) contain *only* values from {-1, -2, -3}?
    is_negative_safe = inc <= {-1, -2, -3}

    # Return True if either condition is met.
    return is_positive_safe or is_negative_safe


# --- Data Input and Preprocessing ---

# This is a single, concise list comprehension to read and format the input file.
data = [
    # Converts strings to integers for one row.
    # 1. x.strip().split(" "): Splits the row string (x) by spaces.
    # 2. if y: Filters out empty strings, which prevents 'ValueError: invalid literal for int()'
    #    when handling multiple spaces between numbers or spaces at the ends of the line.
    # 3. int(y): Converts the remaining strings to integers.
    [int(y) for y in x.strip().split(" ") if y]
    # Reads the file and splits into rows.
    # 1. open(...).read(): Reads the entire file content into a single string.
    # 2. .strip(): Removes leading/trailing whitespace and crucial blank lines from the file content.
    # 3. .split("\n"): Splits the content into a list of individual row strings (x).
    for x in open("../inputs/input-02.txt").read().strip().split("\n")
]

# --- Part 1: Number of Safe Reports - Count of Naturally Safe Rows ---

# sum(...): Treats True as 1 and False as 0, counting the number of rows where the condition is True.
safe_count = sum(
    # A Generator Expression that runs the is_safe() function through every row in the processed 'data'.
    is_safe(row)
    for row in data
)
print(f"Naturaly Safe Rows: {safe_count}")


# --- Part 2: Number of Actual Safe Reports - Count of Rows Made Safe by Removing One Element ---

# The goal is to count how many rows can become 'safe' if *any* single element is removed.
safe_count = sum(
    # An Outer Generator Expression that that runs the is_safe() function through every row in the processed 'data'.
    # The result of this inner part is either True (if a safe sub-sequence was found) or False.
    # An Inner List Comprehension: Generates a sequence of boolean results (True/False) for each removal attempt.
    # row[:i] + row[i + 1 :]: This is list slicing that effectively removes the element at index 'i'.
    #                         It creates a new list from the start up to 'i', and then appends the
    #                         rest of the list from 'i + 1' to the end.
    # any(...): This checks if *at least one* of the generated sequences (the ones with one element removed)
    #           is 'safe'. If any is True, the 'any' function returns True.
    any([is_safe(row[:i] + row[i + 1 :]) for i in range(len(row))])
    for row in data
)
print(f"Actual Safe Rows: {safe_count}")
